### PR TITLE
[dynamic_shapes] SymNode's `hint` does not always conform to `pytype`

### LINKED ISSUE #113393: torch.compile() results in inflexible model with mistralai/Mistral-7B-v0.1
### 🐛 Describe the bug

When applying torch.compile() to HF model `mistralai/Mistral-7B-v0.1`, the resulting model is inflexible in sequence length. The repro code and error message is below:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained(
            "mistralai/Mistral-7B-v0.1",
            torch_dtype=torch.float16,,
            use_cache=True,
        ).to(torch.device('cuda'))

model = torch.compile(model)

#Below three lines work as expected
input_ids =torch.randint(low=0, high=60000, size=(1, 32), dtype=torch.int64)
attention_mask = torch.ones(1, 32, dtype=torch.int64)
model(input_ids=input_ids, attention_mask=attention_mask)

#Next three lines throw error
input_ids =torch.randint(low=0, high=60000, size=(1, 64), dtype=torch.int64)
attention_mask = torch.ones(1, 64, dtype=torch.int64)
model(input_ids=input_ids, attention_mask=attention_mask)

'''
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function not_>(*(Eq(s1, s0) & Eq(s2, s0),), **{}):
The expression should not be static by this point: False
 
from user code:
   File "/dev_data/petermca/python_packages/transformers/models/mistral/modeling_mistral.py", line 295, in forward
    if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
 
Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
 
You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True"
'''
```

### Versions

torch==2.2.0.dev20231108+cu118
transformers==4.34.1

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/113393

Another chapter in the story of Python's horrible handling of int <-> bool interactions.

```python
print(True and 1)  # 1
print(1 and True)  # True
print(True or 1)  # True
print(1 or True)  # 1
```
For sanity's sake, since we have defined more sane type promotion rules, let's use those and ensure `out_hint` conforms to `SymNode`'s `pytype`.

cc @ezyang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/fx/experimental/sym_node.py [hunk 1]
```diff
@@ -95,6 +95,11 @@ def __init__(
         # unbacked symint that a hint was now possible, but as we added more
         # potential refinements to unbacked symints this got harder to keep
         # in sync, so we've deleted it for now.)
+        if hint is not None:
+            assert type(hint) is not pytype, (
+                f"Cannot create SymNode of type {pytype} with "
```
Comment (jon-chuang) [inline torch/fx/experimental/sym_node.py:100]: f-string (and anything after comma) is not evaluated unless assert fails.

**Comment by lezcano** [comment]
Note that these are not int <-> bool interactions but the actual semantics for or and and in Python. In Python, `or` is pretty much equivalent to `return a if bool(a) else b`, and similarly for `and`.

**Comment by jon-chuang** [comment]
Hmm you're right, while it is strictly correct, I wonder if it is better to be stricter as per this PR as I think most people do not write code this way.

But anyway, there is a separate bug here - which is why dynamo is creating sympy values that are int when they are supposed to be bool. I'll try to solve it separately.

**Comment by jon-chuang** [comment]
@lezcano, I am getting some tests which error saying that the hint is of type `SymInt`. To your knowledge, can a hint ever be a `SymInt`, or is this a bug?

cc: @ezyang

**Comment by jon-chuang** [comment]
> @lezcano, I am getting some tests which error saying that the hint is of type SymInt. To your knowledge, can a hint ever be a SymInt?

This seems to only inflict nested and jagged tensors, so I suspect it is a bug.

**Comment by ezyang** [comment]
Hint can be SymInt, this occurs because they use a singleton SymInt to represent jagged dimension. free_symbols will work here instead for the assert

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Actually just test that int accepts either int or SymInt, and same for others

**Pull request closed without merging**